docs/spec_draft.md

```diff
@@ -3408,6 +3409,117 @@ More formally, `results[:][j0, ..., jR-1] = reduce(input_slices)` where:
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.reduce_scatter
+
+### Semantics
+
+<img align="center" src="spec_draft/reduce_scatter.svg" />
+
+Within each process group in the StableHLO grid, performs reduction, using
+`computations`, over the values of the `operand` tensor from each process,
+splits the reduction result along `scatter_dimension` into parts, and scatters
+the split parts between the processes to produce the `result`.
+
+The operation splits the StableHLO grid into `process_groups` as follows:
+  * `channel_id <= 0` and `use_global_device_ids = false`,
+    `cross_replica(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = false`,
+    `cross_replica_and_partition(replica_groups)`.
+  * `channel_id > 0` and `use_global_device_ids = true`,
+    `flattened_ids(replica_groups)`.
+
+Within each `process_group`:
+  * ```
+    reduced_value@process = reduce-all(operand, replica_groups, channel_handle, use_global_device_ids, computation) for each process in process_group
+    ```
+  * ```
+    split_count = dim(process_groups, 1)
+    split_parts@sender = [
+        slice(
+          operand=operand@sender,
+          start_indices=[s0, s1, ..., sR-1],
+            # where
+            #  - sj = 0 if j != split_dimension
+            #  - sj = i * dim(operand) // split_count, if j == split_dimension
+            #  - R = rank(operand)
+          limit_indices=[l0, l1, ..., lR-1],
+            # where
+            #   - lj = dim(operand, j) if j != split_dimension
+            #   - lj = (i + 1) * dim(operand, split_dimension) // split_count, if j == split_dimension
+          strides=[1, ..., 1]
+        ) for i in range(split_count)
+     ]
+     ``` for all `sender` in `process_group`.
+  * `result@receiver = split_parts@sender[receiver_index]` for any sender in process_group,
+      where `receiver_index = index_of(receiver, process_group)`.
+
+### Inputs
+
+| Name                    | Type                                         |
+|-------------------------|----------------------------------------------|
+| `operand`               | tensor of any supported type                 |
+| `scatter_dimension`     | constant of type `si64`                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64` |
+| `channel_id`            | constant of type `si64`                      |
+| `use_global_device_ids` | constant of type `boolean`                   |
+| `computation`           | `function`                                   |
+
+### Outputs
+
+| Name      | Type                                             |
+|-----------|--------------------------------------------------|
+| `result` | tensor of any supported type |
+
+### Constraints
+
+  * (C1) `scatter_dimension` $\in$ [0, rank(`operand`)).
```

Comment:
can we group the constraints on replica_groups, use_global_device_ids etc together at the end and other constraints in the beginning.
i.e. C2, C3, C4, C5

Reply:
We have the constraint s on the result at the end. It follows the order input and then output. 
Adjusted a few constraints. PTAL.